docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,65 @@
+# Title: Password Key Derivation
+
+## Abstract
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not exist.
```

Comment:
this could use more explanation

Reply:
oh, in the following paragraph. okay nevermind

Reply:
> 5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.

?

Reply:
:+1: